tools/shape_tool/ShapeTool.cpp

```diff
@@ -0,0 +1,305 @@
+					tile_window.sprite_tile,
+					tile_window.sprite_palette
+				);
+			}
+		}
+		else
+		{
+			script.g.set_tile(
+				tile_x,
+				tile_y,
+				layer,
+				solid,
+				shape_window.tile_shape,
+				tile_window.sprite_set,
+				tile_window.sprite_tile,
+				tile_window.sprite_palette
+			);
+		}
+	}
+
+	private void pick_tile_at_mouse()
+	{
+		for(int layer=20; layer>=6; --layer)
+		{
+			if(not script.editor.get_layer_visible(layer))
+				continue;
+
+			const float mx = script.g.mouse_x_world(0, layer);
+			const float my = script.g.mouse_y_world(0, layer);
+
+			const int tile_x = int(floor(mx / 48));
+			const int tile_y = int(floor(my / 48));
+
+			tileinfo@ tile = script.g.get_tile(tile_x, tile_y, layer);
+
+			float _;
+			if(tile.solid() and point_in_tile(mx, my, tile_x, tile_y, tile.type(), _, _))
+			{
+				script.editor.set_selected_layer(layer);
+				tile_window.select_tile(tile.sprite_set(), tile.sprite_tile());
+				tile_window.sprite_palette = tile.sprite_palette();
+				shape_window.tile_shape = tile.type();
+				break;
+			}
+		}
+	}
+
+	private void change_layer(int delta)
+	{
+		int layer = max(6, min(20, script.editor.get_selected_layer() + delta));
+		if(layer == 18)
+			layer += sign(delta);
+
+		script.editor.set_selected_layer(layer);
+	}
+
+	private void draw_selection()
+	{
+		const int layer = script.editor.get_selected_layer();
+		int tx1, ty1, tx2, ty2;
+		fix_tile_selection(
+			drag_start_x,
+			drag_start_y,
+			script.g.mouse_x_world(0, layer),
+			script.g.mouse_y_world(0, layer),
+			tx1, ty1, tx2, ty2
+		);
+
+		float x1, y1, x2, y2;
+		script.transform(48 * tx1, 48 * ty1, layer, 10, 19, 10, x1, y1);
+		script.transform(48 * tx2, 48 * ty2, layer, 10, 19, 10, x2, y2);
+
+		script.g.draw_rectangle_world(22, 22, x1, y1, x2, y2, 0, Settings::HoveredFillColour);
+		outline_rect(script.g, 22, 22, x1, y1, x2, y2, 1, Settings::HoveredLineColour);
+	}
+	
+	// //////////////////////////////////////////////////////////
+	// Callbacks
+	// //////////////////////////////////////////////////////////
+	
+	protected void on_select_impl() override
+	{
+		script.editor.hide_panels_gui(true);
+
+		tile_window.show();
+		shape_window.show();
+	}
+	
+	protected void on_deselect_impl() override
+	{
+		script.editor.hide_panels_gui(false);
+
+		tile_window.hide();
+		shape_window.hide();
+	}
+	
+	protected void step_impl() override
+	{
+		if(script.mouse_in_scene and not script.space.down)
+		{
+			if(not shape_window.selecting_tiles)
+			{
+				if(mouse.left_down) set_tile_at_mouse(true);
+				if(mouse.right_down) set_tile_at_mouse(false);
+			}
+
+			if(mouse.middle_press) pick_tile_at_mouse();
+
+			if(script.ctrl.down and mouse.scroll != 0) change_layer(mouse.scroll);
+
+			// Drag start
+			if(mouse.left_press)
+			{
+				dragging = true;
+				const int layer = script.editor.get_selected_layer();
+				drag_start_x = script.g.mouse_x_world(0, layer);
+				drag_start_y = script.g.mouse_y_world(0, layer);
```

Comment:
There is the `mouse_layer` convenience method, which also takes sub layer scale into account since that's a thing now.
https://github.com/cmann1/adv-tools/blob/master/tools/ExtendedTileTool.cpp#L58